File: neurosploit-go/internal/creds/creds.go
package creds

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strings"
	"time"
)

// Login describes an HTTP form-based login flow.
type Login struct {
	URL           string `json:"url"`
	Method        string `json:"method"`
	UsernameField string `json:"username_field"`
	PasswordField string `json:"password_field"`
	Username      string `json:"username"`
	Password      string `json:"password"`
	Success       string `json:"success"`
}

// Ssh holds SSH credentials for Linux host testing.
type Ssh struct {
	Host     string `json:"host"`
	Port     string `json:"port"`
	User     string `json:"user"`
	Password string `json:"password"`
	Key      string `json:"key"`
}

// Win holds Windows / Active Directory credentials.
type Win struct {
	Host     string `json:"host"`
	User     string `json:"user"`
	Password string `json:"password"`
	Domain   string `json:"domain"`
	Hash     string `json:"hash"`
}

// Cloud holds AWS / GCP / Azure credentials for cloud-infra testing.
type Cloud struct {
	AWSAccessKeyID     string `json:"aws_access_key_id,omitempty"`
	AWSSecretAccessKey string `json:"aws_secret_access_key,omitempty"`
	AWSSessionToken    string `json:"aws_session_token,omitempty"`
	AWSRegion          string `json:"aws_region,omitempty"`
	AWSProfile         string `json:"aws_profile,omitempty"`
	GCPServiceAccount  string `json:"gcp_sa_json,omitempty"`
	GCPProject         string `json:"gcp_project,omitempty"`
	AzureTenantID      string `json:"azure_tenant_id,omitempty"`
	AzureClientID      string `json:"azure_client_id,omitempty"`
	AzureClientSecret  string `json:"azure_client_secret,omitempty"`
	AzureSubscription  string `json:"azure_subscription_id,omitempty"`
}

func (c Cloud) isEmpty() bool {
	return c.AWSAccessKeyID == "" && c.AWSProfile == "" &&
		c.GCPServiceAccount == "" && c.AzureClientID == ""
}

// Creds is the loaded credential set from creds.yaml.
type Creds struct {
	JWT    *string `json:"jwt,omitempty"`
	Header *string `json:"header,omitempty"`
	Cookie *string `json:"cookie,omitempty"`
	Login  *Login  `json:"login,omitempty"`
	SSH    *Ssh    `json:"ssh,omitempty"`
	Win    *Win    `json:"win,omitempty"`
	Cloud  *Cloud  `json:"cloud,omitempty"`
}

// Load reads a creds.yaml file and returns the parsed credential set.
// Returns nil if the file is missing or contains no recognizable credential block.
func Load(path string) *Creds {
	text, err := os.ReadFile(path)
	if err != nil {
		return nil
	}
	var c Creds
	login := Login{Method: "POST"}
	ssh := Ssh{Port: "22"}
	win := Win{}
	cloud := Cloud{}
	haveLogin, haveSSH, haveWin := false, false, false
	block := ""
	for _, raw := range strings.Split(string(text), "\n") {
		line := strings.Split(raw, "#")[0]
		if strings.TrimSpace(line) == "" {
			continue
		}
		indented := strings.HasPrefix(line, " ") || strings.HasPrefix(line, "\t")
		idx := strings.Index(line, ":")
		if idx < 0 {
			continue
		}
		k := strings.TrimSpace(line[:idx])
		v := Unquote(strings.TrimSpace(line[idx+1:]))
		if v == "" && !indented {
			switch k {
			case "login":
				haveLogin = true
				block = "login"
			case "ssh":
				haveSSH = true
				block = "ssh"
			case "windows", "win", "ad":
				haveWin = true
				block = "windows"
			case "aws":
				block = "aws"
			case "gcp", "google", "gcloud":
				block = "gcp"
			case "azure", "az":
				block = "azure"
			default:
				block = ""
			}
			continue
		}
		if indented {
			switch block {
			case "login":
				switch k {
				case "url":
					login.URL = v
				case "method":
					login.Method = strings.ToUpper(v)
				case "username_field":
					login.UsernameField = v
				case "password_field":
					login.PasswordField = v
				case "username", "user":
					login.Username = v
				case "password", "pass":
					login.Password = v
				case "success":
					login.Success = v
				}
			case "ssh":
				switch k {
				case "host", "ip":
					ssh.Host = v
				case "port":
					ssh.Port = v
				case "user", "username":
					ssh.User = v
				case "password", "pass":
					ssh.Password = v
				case "key", "keyfile", "identity":
					ssh.Key = v
				}
			case "windows":
				switch k {
				case "host", "ip":
					win.Host = v
				case "user", "username":
					win.User = v
				case "password", "pass":
					win.Password = v
				case "domain":
					win.Domain = v
				case "hash", "ntlm":
					win.Hash = v
				}
			case "aws":
				switch k {
				case "access_key_id", "access_key", "key":
					cloud.AWSAccessKeyID = v
				case "secret_access_key", "secret":
					cloud.AWSSecretAccessKey = v
				case "session_token", "token":
					cloud.AWSSessionToken = v
				case "region":
					cloud.AWSRegion = v
				case "profile":
					cloud.AWSProfile = v
				}
			case "gcp":
				switch k {
				case "service_account_json", "sa_json", "key", "keyfile", "credentials":
					cloud.GCPServiceAccount = v
				case "project", "project_id":
					cloud.GCPProject = v
				}
			case "azure":
				switch k {
				case "tenant_id", "tenant":
					cloud.AzureTenantID = v
				case "client_id", "app_id":
					cloud.AzureClientID = v
				case "client_secret", "secret", "password":
					cloud.AzureClientSecret = v
				case "subscription_id", "subscription":
					cloud.AzureSubscription = v
				}
			}
			continue
		}
		block = ""
		switch k {
		case "jwt", "token":
			c.JWT = &v
		case "header":
			c.Header = &v
		case "cookie":
			c.Cookie = &v
		}
	}
	if haveLogin && login.URL != "" {
		c.Login = &login
	}
	if haveSSH && ssh.Host != "" {
		c.SSH = &ssh
	}
	if haveWin && win.Host != "" {
		c.Win = &win
	}
	if !cloud.isEmpty() {
		c.Cloud = &cloud
	}
	if c.JWT == nil && c.Header == nil && c.Cookie == nil && c.Login == nil && c.SSH == nil && c.Win == nil && c.Cloud == nil {
		return nil
	}
	return &c
}

// AuthHeader returns the request header line to use, or nil if none configured.
func (c *Creds) AuthHeader() *string {
	if c == nil {
		return nil
	}
	if c.Header != nil {
		return c.Header
	}
	if c.JWT != nil {
		s := fmt.Sprintf("Authorization: Bearer %s", *c.JWT)
		return &s
	}
	if c.Cookie != nil {
		s := fmt.Sprintf("Cookie: %s", *c.Cookie)
		return &s
	}
	return nil
}

// CloudEnv returns environment variables for aws/gcloud/az CLIs. Inline GCP JSON is
// written to a temp file and the path is returned as GOOGLE_APPLICATION_CREDENTIALS.
func (c *Creds) CloudEnv() [][2]string {
	if c == nil || c.Cloud == nil {
		return nil
	}
	cl := c.Cloud
	var e [][2]string
	if cl.AWSAccessKeyID != "" {
		e = append(e, [2]string{"AWS_ACCESS_KEY_ID", cl.AWSAccessKeyID})
		e = append(e, [2]string{"AWS_SECRET_ACCESS_KEY", cl.AWSSecretAccessKey})
		if cl.AWSSessionToken != "" {
			e = append(e, [2]string{"AWS_SESSION_TOKEN", cl.AWSSessionToken})
		}
	}
	if cl.AWSProfile != "" {
		e = append(e, [2]string{"AWS_PROFILE", cl.AWSProfile})
	}
	if cl.AWSRegion != "" {
		e = append(e, [2]string{"AWS_DEFAULT_REGION", cl.AWSRegion})
		e = append(e, [2]string{"AWS_REGION", cl.AWSRegion})
	}
	if cl.GCPServiceAccount != "" {
		path := cl.GCPServiceAccount
		if strings.HasPrefix(strings.TrimSpace(path), "{") {
			path = filepath.Join(os.TempDir(), "neurosploit-gcp-sa.json")
			_ = os.WriteFile(path, []byte(cl.GCPServiceAccount), 0600)
		}
		e = append(e, [2]string{"GOOGLE_APPLICATION_CREDENTIALS", path})
	}
	if cl.GCPProject != "" {
		e = append(e, [2]string{"GOOGLE_CLOUD_PROJECT", cl.GCPProject})
		e = append(e, [2]string{"CLOUDSDK_CORE_PROJECT", cl.GCPProject})
	}
	if cl.AzureTenantID != "" {
		e = append(e, [2]string{"AZURE_TENANT_ID", cl.AzureTenantID})
	}
	if cl.AzureClientID != "" {
		e = append(e, [2]string{"AZURE_CLIENT_ID", cl.AzureClientID})
	}
	if cl.AzureClientSecret != "" {
		e = append(e, [2]string{"AZURE_CLIENT_SECRET", cl.AzureClientSecret})
	}
	if cl.AzureSubscription != "" {
		e = append(e, [2]string{"AZURE_SUBSCRIPTION_ID", cl.AzureSubscription})
		e = append(e, [2]string{"ARM_SUBSCRIPTION_ID", cl.AzureSubscription})
	}
	return e
}

// CloudProviderNames returns which cloud providers have credentials configured.
func (c *Creds) CloudProviderNames() []string {
	if c == nil || c.Cloud == nil {
		return nil
	}
	cl := c.Cloud
	var names []string
	if cl.AWSAccessKeyID != "" || cl.AWSProfile != "" {
		names = append(names, "AWS")
	}
	if cl.GCPServiceAccount != "" {
		names = append(names, "GCP")
	}
	if cl.AzureClientID != "" {
		names = append(names, "Azure")
	}
	return names
}

// CloudInstruction returns a directive describing available cloud credentials.
func (c *Creds) CloudInstruction() *string {
	if c == nil || c.Cloud == nil {
		return nil
	}
	cl := c.Cloud
	var s strings.Builder
	if cl.AWSAccessKeyID != "" || cl.AWSProfile != "" {
		region := ""
		if cl.AWSRegion != "" {
			region = fmt.Sprintf(" (region %s)", cl.AWSRegion)
		}
		fmt.Fprintf(&s, "AWS ACCESS: credentials are set in the environment%s. Use the `aws` CLI to enumerate and test the account — start with `aws sts get-caller-identity`, then IAM (users/roles/policies, privilege escalation paths), S3 (public/misconfigured buckets), EC2/SG, Lambda, Secrets Manager. Read-only enumeration first; never destructive.\n", region)
	}
	if cl.GCPServiceAccount != "" {
		project := ""
		if cl.GCPProject != "" {
			project = fmt.Sprintf(" (project %s)", cl.GCPProject)
		}
		fmt.Fprintf(&s, "GCP ACCESS: a service account is available via $GOOGLE_APPLICATION_CREDENTIALS%s. Run `gcloud auth activate-service-account --key-file=$GOOGLE_APPLICATION_CREDENTIALS` first, then enumerate with `gcloud`/`gsutil` — IAM bindings & privilege escalation, buckets, compute, service accounts/keys, Cloud Functions.\n", project)
	}
	if cl.AzureClientID != "" {
		s.WriteString("AZURE ACCESS: a service principal is set in the environment. Authenticate with `az login --service-principal -u $AZURE_CLIENT_ID -p $AZURE_CLIENT_SECRET --tenant $AZURE_TENANT_ID`, then enumerate with `az` — role assignments (RBAC) & escalation, storage accounts/containers, VMs, Key Vaults, managed identities.\n")
	}
	if s.Len() == 0 {
		return nil
	}
	out := s.String()
	return &out
}

// HostInstruction returns a directive describing host credentials available to agents.
func (c *Creds) HostInstruction() *string {
	if c == nil {
		return nil
	}
	var s strings.Builder
	if c.SSH != nil {
		auth := "password (provided)"
		if c.SSH.Key != "" {
			auth = fmt.Sprintf("private key %s", c.SSH.Key)
		}
		fmt.Fprintf(&s,
			"SSH ACCESS (Linux): host %s:%s as user '%s' via %s. Use `ssh`/`sshpass` to run enumeration and privilege-escalation checks on the host.\n",
			c.SSH.Host, c.SSH.Port, c.SSH.User, auth)
	}
	if c.Win != nil {
		auth := "password"
		if c.Win.Hash != "" {
			auth = "NTLM hash (pass-the-hash)"
		}
		domain := "(workgroup)"
		if c.Win.Domain != "" {
			domain = c.Win.Domain
		}
		fmt.Fprintf(&s,
			"WINDOWS/AD ACCESS: host %s domain '%s' as user '%s' via %s. Use tools like crackmapexec/netexec, impacket, evil-winrm, bloodhound-python for host and AD checks.\n",
			c.Win.Host, domain, c.Win.User, auth)
	}
	if s.Len() == 0 {
		return nil
	}
	out := s.String()
	return &out
}

// LoginInstruction returns a directive instructing the agent to authenticate first.
func (c *Creds) LoginInstruction() *string {
	if c == nil || c.Login == nil {
		return nil
	}
	l := c.Login
	s := fmt.Sprintf("AUTHENTICATE FIRST: %s %s with %s=%s and %s=%s; capture the session cookie/token from the response (success indicator: \"%s\") and reuse it on every subsequent request.",
		l.Method, l.URL, l.UsernameField, l.Username, l.PasswordField, l.Password, l.Success)
	return &s
}

// Unquote strips matching surrounding quotes from a value.
func Unquote(s string) string {
	s = strings.TrimSpace(s)
	if len(s) >= 2 {
		if (s[0] == '"' && s[len(s)-1] == '"') || (s[0] == '\'' && s[len(s)-1] == '\'') {
			return s[1 : len(s)-1]
		}
	}
	return s
}

// DoLogin performs the configured HTTP login flow and returns an auth header to reuse.
func DoLogin(ctx context.Context, l *Login) (string, string, error) {
	if l == nil || l.URL == "" {
		return "", "", fmt.Errorf("no login configured")
	}
	client := &http.Client{
		Timeout: 30 * time.Second,
		CheckRedirect: func(req *http.Request, via []*http.Request) error {
			return http.ErrUseLastResponse
		},
	}

	form := url.Values{}
	form.Set(l.UsernameField, l.Username)
	form.Set(l.PasswordField, l.Password)

	var req *http.Request
	var err error
	if strings.ToUpper(l.Method) == "GET" {
		u, parseErr := url.Parse(l.URL)
		if parseErr != nil {
			return "", "", fmt.Errorf("parse login URL: %w", parseErr)
		}
		q := u.Query()
		for k, v := range form {
			for _, vv := range v {
				q.Add(k, vv)
			}
		}
		u.RawQuery = q.Encode()
		req, err = http.NewRequestWithContext(ctx, http.MethodGet, u.String(), nil)
	} else {
		req, err = http.NewRequestWithContext(ctx, http.MethodPost, l.URL, strings.NewReader(form.Encode()))
		if err == nil {
			req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
		}
	}
	if err != nil {
		return "", "", fmt.Errorf("build request: %w", err)
	}

	resp, err := client.Do(req)
	if err != nil {
		return "", "", fmt.Errorf("login request: %w", err)
	}
	defer func() { _ = resp.Body.Close() }()

	status := resp.StatusCode
	var cookiePairs []string
	for _, c := range resp.Cookies() {
		if c != nil && c.Name != "" {
			cookiePairs = append(cookiePairs, fmt.Sprintf("%s=%s", c.Name, c.Value))
		}
	}
	body, _ := io.ReadAll(resp.Body)

	var bodyObj map[string]interface{}
	if err := json.Unmarshal(body, &bodyObj); err == nil {
		for _, k := range []string{"access_token", "token", "jwt", "id_token", "accessToken"} {
			if v, ok := bodyObj[k]; ok {
				if s, ok := v.(string); ok && s != "" {
					return fmt.Sprintf("Authorization: Bearer %s", s), fmt.Sprintf("bearer token from JSON `%s` (HTTP %d)", k, status), nil
				}
			}
		}
	}

	if len(cookiePairs) > 0 {
		cookie := strings.Join(cookiePairs, "; ")
		ok := l.Success == "" || strings.Contains(string(body), l.Success) || status == http.StatusFound || status == http.StatusMovedPermanently || (status >= 200 && status < 300)
		extra := ""
		if !ok {
			extra = ", success marker not seen"
		}
		return fmt.Sprintf("Cookie: %s", cookie), fmt.Sprintf("session cookie captured (HTTP %d%s)", status, extra), nil
	}
	return "", "", fmt.Errorf("login returned no Set-Cookie or token (HTTP %d)", status)
}

File: neurosploit-rs/crates/harness/src/creds.rs
//! Credential loading for authenticated testing (`creds.yaml`).
//!
//! Dependency-free parser for a small YAML subset: flat `key: value` pairs plus
//! one nested `login:` block (2-space indent). Lets the operator hand the
//! harness a JWT / header / cookie, or a login flow the agents should perform so
//! they test the target as an authenticated user.
//!
//! Example `creds.yaml`:
//! ```yaml
//! jwt: eyJhbGciOi...                 # → Authorization: Bearer <jwt>
//! # header: "X-Api-Key: abc123"      # raw header (alternative)
//! # cookie: "session=deadbeef"       # → Cookie: session=deadbeef
//! login:
//!   url: http://app/login
//!   method: POST
//!   username_field: uid
//!   password_field: passw
//!   username: admin
//!   password: admin
//!   success: Logout
//! ```

#[derive(Default, Debug, Clone)]
pub struct Login {
    pub url: String,
    pub method: String,
    pub username_field: String,
    pub password_field: String,
    pub username: String,
    pub password: String,
    pub success: String,
}

/// SSH credentials for Linux host testing.
#[derive(Default, Debug, Clone)]
pub struct Ssh {
    pub host: String,
    pub port: String,   // default 22
    pub user: String,
    pub password: String,
    pub key: String,    // path to a private key
}

/// Windows / Active Directory credentials.
#[derive(Default, Debug, Clone)]
pub struct Win {
    pub host: String,
    pub user: String,
    pub password: String,
    pub domain: String,
    pub hash: String,   // NTLM hash for pass-the-hash (LM:NT or NT)
}

/// Cloud provider credentials for cloud-infra testing (AWS / GCP / Azure).
/// Secrets are read from `creds.yaml` and exported to the process environment so
/// the `aws` / `gcloud` / `az` CLIs the agents use pick them up automatically.
#[derive(Default, Debug, Clone)]
pub struct Cloud {
    // AWS — static keys (access key + secret [+ session token]) OR a named profile.
    pub aws_access_key_id: String,
    pub aws_secret_access_key: String,
    pub aws_session_token: String,
    pub aws_region: String,
    pub aws_profile: String,
    // GCP — a service-account JSON (path, recommended) or inline single-line JSON.
    pub gcp_sa_json: String,
    pub gcp_project: String,
    // Azure — a service principal (recommended for non-interactive automation).
    pub azure_tenant_id: String,
    pub azure_client_id: String,
    pub azure_client_secret: String,
    pub azure_subscription_id: String,
}

impl Cloud {
    fn is_empty(&self) -> bool {
        self.aws_access_key_id.is_empty() && self.aws_profile.is_empty()
            && self.gcp_sa_json.is_empty()
            && self.azure_client_id.is_empty()
    }
}

#[derive(Default, Debug, Clone)]
pub struct Creds {
    pub jwt: Option<String>,
    pub header: Option<String>,
    pub cookie: Option<String>,
    pub login: Option<Login>,
    pub ssh: Option<Ssh>,
    pub win: Option<Win>,
    pub cloud: Option<Cloud>,
}

impl Creds {
    pub fn load(path: &std::path::Path) -> Option<Creds> {
        let text = std::fs::read_to_string(path).ok()?;
        let mut c = Creds::default();
        let mut login = Login { method: "POST".into(), ..Default::default() };
        let mut ssh = Ssh { port: "22".into(), ..Default::default() };
        let mut win = Win::default();
        let mut cloud = Cloud::default();
        let (mut have_login, mut have_ssh, mut have_win) = (false, false, false);
        let mut block = ""; // "", "login", "ssh", "windows", "aws", "gcp", "azure"
        for raw in text.lines() {
            let line = raw.split('#').next().unwrap_or("");
            if line.trim().is_empty() {
                continue;
            }
            let indented = line.starts_with(' ') || line.starts_with('\t');
            let (k, v) = match line.split_once(':') {
                Some((k, v)) => (k.trim().to_string(), unquote(v.trim())),
                None => continue,
            };
            // Enter a nested block (header line with empty value).
            if v.is_empty() && !indented {
                block = match k.as_str() {
                    "login" => { have_login = true; "login" }
                    "ssh" => { have_ssh = true; "ssh" }
                    "windows" | "win" | "ad" => { have_win = true; "windows" }
                    "aws" => "aws",
                    "gcp" | "google" | "gcloud" => "gcp",
                    "azure" | "az" => "azure",
                    _ => "",
                };
                continue;
            }
            if indented {
                match block {
                    "login" => match k.as_str() {
                        "url" => login.url = v,
                        "method" => login.method = v.to_uppercase(),
                        "username_field" => login.username_field = v,
                        "password_field" => login.password_field = v,
                        "username" | "user" => login.username = v,
                        "password" | "pass" => login.password = v,
                        "success" => login.success = v,
                        _ => {}
                    },
                    "ssh" => match k.as_str() {
                        "host" | "ip" => ssh.host = v,
                        "port" => ssh.port = v,
                        "user" | "username" => ssh.user = v,
                        "password" | "pass" => ssh.password = v,
                        "key" | "keyfile" | "identity" => ssh.key = v,
                        _ => {}
                    },
                    "windows" => match k.as_str() {
                        "host" | "ip" => win.host = v,
                        "user" | "username" => win.user = v,
                        "password" | "pass" => win.password = v,
                        "domain" => win.domain = v,
                        "hash" | "ntlm" => win.hash = v,
                        _ => {}
                    },
                    "aws" => match k.as_str() {
                        "access_key_id" | "access_key" | "key" => cloud.aws_access_key_id = v,
                        "secret_access_key" | "secret" => cloud.aws_secret_access_key = v,
                        "session_token" | "token" => cloud.aws_session_token = v,
                        "region" => cloud.aws_region = v,
                        "profile" => cloud.aws_profile = v,
                        _ => {}
                    },
                    "gcp" => match k.as_str() {
                        "service_account_json" | "sa_json" | "key" | "keyfile" | "credentials" => cloud.gcp_sa_json = v,
                        "project" | "project_id" => cloud.gcp_project = v,
                        _ => {}
                    },
                    "azure" => match k.as_str() {
                        "tenant_id" | "tenant" => cloud.azure_tenant_id = v,
                        "client_id" | "app_id" => cloud.azure_client_id = v,
                        "client_secret" | "secret" | "password" => cloud.azure_client_secret = v,
                        "subscription_id" | "subscription" => cloud.azure_subscription_id = v,
                        _ => {}
                    },
                    _ => {}
                }
                continue;
            }
            block = "";
            match k.as_str() {
                "jwt" | "token" => c.jwt = Some(v),
                "header" => c.header = Some(v),
                "cookie" => c.cookie = Some(v),
                _ => {}
            }
        }
        if have_login && !login.url.is_empty() { c.login = Some(login); }
        if have_ssh && !ssh.host.is_empty() { c.ssh = Some(ssh); }
        if have_win && !win.host.is_empty() { c.win = Some(win); }
        if !cloud.is_empty() { c.cloud = Some(cloud); }
        if c.jwt.is_none() && c.header.is_none() && c.cookie.is_none()
            && c.login.is_none() && c.ssh.is_none() && c.win.is_none() && c.cloud.is_none() {
            return None;
        }
        Some(c)
    }

    /// Environment variables to export so the `aws`/`gcloud`/`az` CLIs the agents
    /// run pick up the cloud credentials automatically. For inline GCP JSON the
    /// content is written to a temp file and that path is returned.
    pub fn cloud_env(&self) -> Vec<(String, String)> {
        let mut e: Vec<(String, String)> = Vec::new();
        let Some(c) = &self.cloud else { return e };
        // AWS
        if !c.aws_access_key_id.is_empty() {
            e.push(("AWS_ACCESS_KEY_ID".into(), c.aws_access_key_id.clone()));
            e.push(("AWS_SECRET_ACCESS_KEY".into(), c.aws_secret_access_key.clone()));
            if !c.aws_session_token.is_empty() {
                e.push(("AWS_SESSION_TOKEN".into(), c.aws_session_token.clone()));
            }
        }
        if !c.aws_profile.is_empty() { e.push(("AWS_PROFILE".into(), c.aws_profile.clone())); }
        if !c.aws_region.is_empty() {
            e.push(("AWS_DEFAULT_REGION".into(), c.aws_region.clone()));
            e.push(("AWS_REGION".into(), c.aws_region.clone()));
        }
        // GCP — path (recommended) or inline JSON written to a temp file.
        if !c.gcp_sa_json.is_empty() {
            let path = if c.gcp_sa_json.trim_start().starts_with('{') {
                let p = std::env::temp_dir().join("neurosploit-gcp-sa.json");
                let _ = std::fs::write(&p, c.gcp_sa_json.as_bytes());
                p.display().to_string()
            } else {
                c.gcp_sa_json.clone()
            };
            e.push(("GOOGLE_APPLICATION_CREDENTIALS".into(), path));
        }
        if !c.gcp_project.is_empty() {
            e.push(("GOOGLE_CLOUD_PROJECT".into(), c.gcp_project.clone()));
            e.push(("CLOUDSDK_CORE_PROJECT".into(), c.gcp_project.clone()));
        }
        // Azure — service principal env (consumed by `az login --service-principal`).
        if !c.azure_tenant_id.is_empty() { e.push(("AZURE_TENANT_ID".into(), c.azure_tenant_id.clone())); }
        if !c.azure_client_id.is_empty() { e.push(("AZURE_CLIENT_ID".into(), c.azure_client_id.clone())); }
        if !c.azure_client_secret.is_empty() { e.push(("AZURE_CLIENT_SECRET".into(), c.azure_client_secret.clone())); }
        if !c.azure_subscription_id.is_empty() {
            e.push(("AZURE_SUBSCRIPTION_ID".into(), c.azure_subscription_id.clone()));
            e.push(("ARM_SUBSCRIPTION_ID".into(), c.azure_subscription_id.clone()));
        }
        e
    }

    /// A directive telling the agents which cloud creds are available and how to
    /// authenticate the provider CLI, so they enumerate/test the cloud account.
    pub fn cloud_instruction(&self) -> Option<String> {
        let c = self.cloud.as_ref()?;
        let mut s = String::new();
        if !c.aws_access_key_id.is_empty() || !c.aws_profile.is_empty() {
            s.push_str(&format!(
                "AWS ACCESS: credentials are set in the environment{}. Use the `aws` CLI to enumerate and test the account — start with `aws sts get-caller-identity`, then IAM (users/roles/policies, privilege escalation paths), S3 (public/misconfigured buckets), EC2/SG, Lambda, Secrets Manager. Read-only enumeration first; never destructive.\n",
                if c.aws_region.is_empty() { String::new() } else { format!(" (region {})", c.aws_region) }));
        }
        if !c.gcp_sa_json.is_empty() {
            s.push_str(&format!(
                "GCP ACCESS: a service account is available via $GOOGLE_APPLICATION_CREDENTIALS{}. Run `gcloud auth activate-service-account --key-file=$GOOGLE_APPLICATION_CREDENTIALS` first, then enumerate with `gcloud`/`gsutil` — IAM bindings & privilege escalation, buckets, compute, service accounts/keys, Cloud Functions.\n",
                if c.gcp_project.is_empty() { String::new() } else { format!(" (project {})", c.gcp_project) }));
        }
        if !c.azure_client_id.is_empty() {
            s.push_str(
                "AZURE ACCESS: a service principal is set in the environment. Authenticate with `az login --service-principal -u $AZURE_CLIENT_ID -p $AZURE_CLIENT_SECRET --tenant $AZURE_TENANT_ID`, then enumerate with `az` — role assignments (RBAC) & escalation, storage accounts/containers, VMs, Key Vaults, managed identities.\n");
        }
        if s.is_empty() { None } else { Some(s) }
    }

    /// A directive describing the host credentials available to the agents, so
    /// they can authenticate to Linux (SSH) / Windows (AD) hosts.
    pub fn host_instruction(&self) -> Option<String> {
        let mut s = String::new();
        if let Some(h) = &self.ssh {
            let auth = if !h.key.is_empty() { format!("private key {}", h.key) } else { "password (provided)".into() };
            s.push_str(&format!(
                "SSH ACCESS (Linux): host {}:{} as user '{}' via {}. Use `ssh`/`sshpass` to run \
                 enumeration and privilege-escalation checks on the host.\n",
                h.host, h.port, h.user, auth));
        }
        if let Some(w) = &self.win {
            let auth = if !w.hash.is_empty() { "NTLM hash (pass-the-hash)".to_string() } else { "password".into() };
            s.push_str(&format!(
                "WINDOWS/AD ACCESS: host {} domain '{}' as user '{}' via {}. Use tools like \
                 crackmapexec/netexec, impacket, evil-winrm, bloodhound-python for host and AD checks.\n",
                w.host, if w.domain.is_empty() { "(workgroup)" } else { &w.domain }, w.user, auth));
        }
        if s.is_empty() { None } else { Some(s) }
    }

    /// The auth material to send with each request, as a header line.
    pub fn auth_header(&self) -> Option<String> {
        if let Some(h) = &self.header {
            return Some(h.clone());
        }
        if let Some(j) = &self.jwt {
            return Some(format!("Authorization: Bearer {j}"));
        }
        if let Some(ck) = &self.cookie {
            return Some(format!("Cookie: {ck}"));
        }
        None
    }

    /// A directive instructing the agent to authenticate first via curl.
    pub fn login_instruction(&self) -> Option<String> {
        let l = self.login.as_ref()?;
        Some(format!(
            "AUTHENTICATE FIRST: {} {} with {}={} and {}={}; capture the session cookie/token \
             from the response (success indicator: \"{}\") and reuse it on every subsequent request.",
            l.method, l.url, l.username_field, l.username, l.password_field, l.password, l.success
        ))
    }
}

/// Perform the login flow now (real HTTP POST) and return an auth header to
/// reuse on every subsequent request: a `Cookie:` from Set-Cookie, or an
/// `Authorization: Bearer` from a token in the JSON response. Returns
/// (auth_header, note). Redirects are not followed so the login response's
/// Set-Cookie is visible.
pub async fn login(l: &Login) -> anyhow::Result<(String, String)> {
    use reqwest::header::SET_COOKIE;
    let client = reqwest::Client::builder()
        .redirect(reqwest::redirect::Policy::none())
        .timeout(std::time::Duration::from_secs(30))
        .build()?;
    let form: Vec<(String, String)> = vec![
        (l.username_field.clone(), l.username.clone()),
        (l.password_field.clone(), l.password.clone()),
    ];
    let req = if l.method == "GET" {
        client.get(&l.url).query(&form)
    } else {
        client.post(&l.url).form(&form)
    };
    let resp = req.send().await?;
    let status = resp.status();

    // 1) session cookies from Set-Cookie on the login response
    let mut cookie_pairs = Vec::new();
    for hv in resp.headers().get_all(SET_COOKIE) {
        if let Ok(s) = hv.to_str() {
            if let Some(pair) = s.split(';').next() {
                let p = pair.trim();
                if !p.is_empty() {
                    cookie_pairs.push(p.to_string());
                }
            }
        }
    }
    let body = resp.text().await.unwrap_or_default();

    // 2) bearer token from a JSON response body
    if let Ok(v) = serde_json::from_str::<serde_json::Value>(&body) {
        for k in ["access_token", "token", "jwt", "id_token", "accessToken"] {
            if let Some(t) = v.get(k).and_then(|x| x.as_str()).filter(|t| !t.is_empty()) {
                return Ok((format!("Authorization: Bearer {t}"), format!("bearer token from JSON `{k}` (HTTP {status})")));
            }
        }
    }
    if !cookie_pairs.is_empty() {
        let cookie = cookie_pairs.join("; ");
        // Soft success check (don't fail hard — many apps 302 on success).
        let ok = l.success.is_empty() || body.contains(&l.success) || status.is_redirection() || status.is_success();
        let note = format!("session cookie captured (HTTP {status}{})", if ok { "" } else { ", success marker not seen" });
        return Ok((format!("Cookie: {cookie}"), note));
    }
    anyhow::bail!("login returned no Set-Cookie or token (HTTP {status})")
}

fn unquote(s: &str) -> String {
    let s = s.trim();
    if (s.starts_with('"') && s.ends_with('"') && s.len() >= 2)
        || (s.starts_with('\'') && s.ends_with('\'') && s.len() >= 2)
    {
        s[1..s.len() - 1].to_string()
    } else {
        s.to_string()
    }
}
